### PR TITLE
[SPARK-26138][SQL] Cross join requires push LocalLimit in LimitPushDown rule

### PR DESCRIPTION
**Pull request by guoxiaolongzte**
## What changes were proposed in this pull request?

In LimitPushDown batch, cross join can push down the limit.

## How was this patch tested?

manual tests

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -459,6 +459,7 @@ object LimitPushDown extends Rule[LogicalPlan] {
       val newJoin = joinType match {
         case RightOuter => join.copy(right = maybePushLocalLimit(exp, right))
         case LeftOuter => join.copy(left = maybePushLocalLimit(exp, left))
+        case Cross => join.copy(left = maybePushLocalLimit(exp, left))
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: what if the other table is empty?

**Review by mgaido91 — CHANGES_REQUESTED** [review]
Please add some UTs in order to enforce this. Moreover, can't we push it on the right side as well?

**Comment by guoxiaolongzte** [comment]
OK,  I will add some UTs.

**Comment by guoxiaolongzte** [comment]
Cartesian product refers to the Cartesian product of two sets X and Y in mathematics , also known as direct product , expressed as X × Y , the first object is a member of X and the second object is One of all possible ordered pairs of Y.So cross join mustpush it on the left side.
   For example, A={a,b}, B={0,1,2}, then
   A × B = {(a, 0), (a, 1), (a, 2), (b, 0), (b, 1), (b, 2)}
   B×A={(0, a), (0, b), (1, a), (1, b), (2, a), (2, b)}

**Comment by mgaido91** [comment]
@guoxiaolongzte still that doesn't explain why we can push to the right side too. I do believe that it is possible. If the right side contains more than N items, where N is the limit size, the output will contains the combinations of the first item from the left side and the first N items from the right side. If the right side contains less than N items, pushing the limit on its side has no effect on the result.

**Comment by guoxiaolongzte** [comment]
Yes I tested and understood, you are right. @mgaido91

**Review by mgaido91 — COMMENTED** [review]
now this seems reasonable to me. cc @cloud-fan @dongjoon-hyun @gatorsmile shall we trigger a build for this? Thanks.

**Comment by guoxiaolongzte** [comment]
@cloud-fan @dongjoon-hyun @gatorsmile
Help review the code.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -459,6 +459,7 @@ object LimitPushDown extends Rule[LogicalPlan] {
       val newJoin = joinType match {
         case RightOuter => join.copy(right = maybePushLocalLimit(exp, right))
         case LeftOuter => join.copy(left = maybePushLocalLimit(exp, left))
+        case Cross => join.copy(left = maybePushLocalLimit(exp, left), right = maybePushLocalLimit(exp, right))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: how about inner join without condition?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: maybe we  can match `InnerLike` when condition is empty.

Comment (guoxiaolongzte) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: A = {(a, 0), (b, 1), (c, 2), (d, 0), **(e, 1), (f, 2)**}
B = {**(e, 1), (f, 2)**}

A inner join B limit 2
If there is limit 2, (a, 0), (b, 1) inner join {(e, 1), (f, 2)}, the result is empty. But the real result is not empty.

Comment (guoxiaolongzte) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: When set  spark.sql.crossJoin.enabled=true, 
inner join without condition, LeftOuter without condition, RightOuter without condition, FullOuter without condition， all these are  iterally cross join?

Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: +1

**Comment by gatorsmile** [comment]
The title has a typo.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 2 above)
Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: @guoxiaolongzte nope.

**Comment by guoxiaolongzte** [comment]
> The title has a typo.

Sorry, it has been fixed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 2 above)
Comment (guoxiaolongzte) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: @cloud-fan 
Please give me some advice. Thank you.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: if there is no join condition, I think join type doesn't matter and we can always push down limits.

Comment (guoxiaolongzte) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: There are two tables as follows：
CREATE TABLE `**test1**`(`id` int, `name` int);
CREATE TABLE `**test2**`(`id` int, `name` int);

test1 table data:
2,2
1,1

test2 table data:
2,2
3,3
4,4

Execute sql select * from test1 t1 **left anti join** test2 t2 on t1.id=t2.id limit 1; The result:
1,1

But 
   we push the limit 1 on left side, the result is not correct. Result is empty.
   we push the limit 1 on right side, the result is not correct. Result is empty.

So
    left anti join no need to push down limit. Similarly, left semi join is the same logic.

**Comment by liu-zhaokun** [comment]
@guoxiaolongzte  good job

**Comment by guoxiaolongzte** [comment]
Can I give you some advice on this issue?@gatorsmile @cloud-fan

**Comment by dongjoon-hyun** [comment]
ok to test.

**Comment by dongjoon-hyun** [comment]
Hi, @guoxiaolongzte . Could you run `dev/scalastyle` and fix the issue?

**Comment by guoxiaolongzte** [comment]
@dongjoon-hyun 
Thank you very much for your review.

**Comment by guoxiaolongzte** [comment]
I looked at it. This error is not caused by my pr.

**Comment by mgaido91** [comment]
@guoxiaolongzte can you address @cloud-fan 's comment? We need the same for InnerLike joins without conditions...

**Pull request closed without merging**